src/sqlfluff/rules/L042.py

```diff
@@ -48,44 +70,281 @@ class Rule_L042(BaseRule):
-
-                # Is it bracketed? If so, lint that instead.
-                bracketed_expression = from_expression_element.children(
-                    is_type("bracketed")
+    def _eval(self, context: RuleContext) -> Optional[List[LintResult]]:
+        """Join/From clauses should not contain subqueries. Use CTEs instead."""
+        select_types = [
+            "with_compound_statement",
+            "set_expression",
+            "select_statement",
+        ]
+        self.forbid_subquery_in: str
+        parent_types = self._config_mapping[self.forbid_subquery_in]
+        segment = context.functional.segment
+        parent_stack = context.functional.parent_stack
+        is_select = segment.all(is_type(*select_types))
+        is_select_child = parent_stack.any(is_type(*select_types))
+        if not is_select or is_select_child:
+            # Subvert the Crawler
+            return None
+
+        root_select = segment
+        nested_subqueries: List[Tuple[str, Segments, Segments, BaseSegment]] = []
+        selects = segment.recursive_crawl(*select_types, recurse_into=True)
+        for select in selects.iterate_segments():
+            for res in _find_nested_subqueries(select):
+                clause_type = res[0]
+                if clause_type not in parent_types:
+                    continue
+                nested_subqueries.append(res)
+
+        if not nested_subqueries:
+            return None
+
+        return _calculate_fixes(context.dialect, root_select, nested_subqueries)
+
+
+S = TypeVar("S", bound=Type[BaseSegment])
+
+
+def _get_seg(class_def: S, dialect: Dialect) -> S:
+    return cast(S, dialect.get_segment(class_def.__name__))
+
+
+def _calculate_fixes(
+    dialect: Dialect,
+    root_select: Segments,
+    nested_subqueries: List[Tuple[str, Segments, Segments, BaseSegment]],
+):
+    """Given the Root select and the offending subqueries calculate fixes."""
+    Seg = partial(_get_seg, dialect=dialect)
+    ctes = _CTEChecker()
+    is_with = root_select.all(is_type("with_compound_statement"))
+    # TODO: consider if we can fix recursive CTEs
+    is_recursive = is_with and len(root_select.children(is_name("recursive"))) > 0
+    segmentify = partial(
+        _segmentify,
+        casing=_get_casing_preference(root_select),
+    )
+    # Init the output/final select &
+    # populate existing CTEs
+    output_select = root_select
+    if is_with:
+        output_select = root_select.children(
+            is_type(
+                "set_expression",
+                "select_statement",
+            )
+        )
+        for cte in root_select.children(is_type("common_table_expression")):
+            assert isinstance(cte, CTEDefinitionSegment), "TypeGaurd"
+            ctes.insert_cte(cte)
+
+    lint_results: List[LintResult] = []
+    CTESegment = Seg(CTEDefinitionSegment)
+    TableExpressionSeg = Seg(TableExpressionSegment)
+    TableReferenceSeg = Seg(TableReferenceSegment)
```

Comment:
Here I felt the issue again, of not being able to just lex & pars the new desired input. `change[str] --> Segments`

Basically TableExpressionSeg and TableReferenceSegment have been subclassed based on Dialect. I was concerned about strange behaviour when adding a segment explicitly using the ansi class when the query being processed is using/expecting the BigQuery version of the class everywhere.

Reply:
It would be cool to do this. Interested in taking a look at that? I may have time, maybe not. (I'm starting a new job next week. We'll see how it affects my schedule.)

Reply:
I will try if I get a chance :)

Reply:
I think there are some existing parser or dialect tests that already do this in isolation, so it may be as easy as calling those same functions on context.dialect. 🤞